compute_sdk/globus_compute_sdk/sdk/executor.py

```diff
@@ -191,37 +202,42 @@ def __init__(
 
     def __repr__(self) -> str:
         name = self.__class__.__name__
-        label = self.label and f"{self.label}, " or ""
-        ep_id = self.endpoint_id and f", ep_id:{self.endpoint_id}" or ""
-        c_id = self.container_id and f", c_id:{self.container_id}" or ""
-        tg_id = f", tg_id:{self.task_group_id}"
-        return f"{name}<{label}{self.batch_size}{ep_id}{c_id}{tg_id}>"
+        label = self.label and f"{self.label}; " or ""
+        ep_id = self.endpoint_id and f"ep_id:{self.endpoint_id}; " or ""
+        c_id = self.container_id and f"c_id:{self.container_id}; " or ""
+        tg_id = f"tg_id:{self.task_group_id}; "
+        bs = f"bs:{self.batch_size}"
+        return f"{name}<{label}{ep_id}{c_id}{tg_id}{bs}>"
 
     @property
-    def task_group_id(self) -> str:
+    def task_group_id(self) -> UUID_LIKE_T | None:
         """
         The Task Group with which this instance is currently associated.  New tasks will
         be sent to this Task Group upstream, and the result listener will only listen
         for results for this group.
 
-        Must be a string.  Set by simple assignment::
+        Must be a valid UUID.  Set by simple assignment::
 
             gce = Executor(endpoint_id="...")
-            gce.task_group_id = "Some-stored-id"
+            gce.task_group_id = "00000000-0000-0000-0000-000000000000"
 
         This is typically used when reattaching to a previously initiated set of tasks.
         See `reload_tasks()`_ for more information.
 
-        [default: ``None``, which translates to the Client task group id]
+        If not set manually, this will be set automatically on `submit()`_, to a Task
+        Group ID supplied by the services.
+
+        [default: ``None``]
         """
         return self._task_group_id
 
     @task_group_id.setter
-    def task_group_id(self, task_group_id: str):
+    def task_group_id(self, task_group_id: UUID_LIKE_T | None):
         self._task_group_id = task_group_id
 
     def _fn_cache_key(self, fn: t.Callable):
-        return (fn, self.container_id)
+        c_id = str(self.container_id) if self.container_id else None
+        return (fn, c_id)
```

Comment:
This change took a minute for me to grok: because now `container_id` is `UUID_LIKE_T | None` rather than `str | None`.  Okay, makes sense.  On the other hand, that enters code smell for me: there should be one canonical representation, and, where feasible and losslessly equal, it should be the "less expensive" one (e.g., memory, or CPU, or ...).

Suggestion: let's _not_ address this smell in this PR&nbsp;&mdash;&nbsp;this PR has iterated enough and is very nearly there&nbsp;&mdash;&nbsp;but let's create a ticket to fix this up later.  The fix will be simple, I think, just flushing through that `container_id` _also_ is a property, and in the setter, convert to a UUID (or `str`!  Do the analysis for which is better!) or `None`.  Something like:

```python
@property
def container_id(self) -> UUID_LIKE_T:
    """description here"""
    return self._task_group_id

@container_id.setter
def container_id(self, c_id: UUID_LIKE_T | None):
    if c_id:
        c_id = c_id if isinstance(c_id, uuid.UUID) else uuid.UUID(c_id)
    self._container_id = c_id
```

(And with ticket created, potentially link that TODO here.)